src/audio/tdfb/tdfb.c

```diff
@@ -57,10 +63,58 @@ DECLARE_SOF_RT_UUID("tdfb", tdfb_uuid,  0xdd511749, 0xd9fa, 0x455c, 0xb3, 0xa7,
 
 DECLARE_TR_CTX(tdfb_tr, SOF_UUID(tdfb_uuid), LOG_LEVEL_INFO);
 
-/* IPC */
+/* IPC helpers for control update to user space */
 
-/* TODO: ALSA control update to user space need to be moved to module adapter */
+#if CONFIG_IPC_MAJOR_4
+static struct ipc_msg *tdfb_notification_init(struct processing_module *mod,
+					      uint32_t control_type_param_id,
+					      uint32_t control_id)
+{
+	struct ipc_msg msg_proto;
+	struct comp_dev *dev = mod->dev;
+	struct comp_ipc_config *ipc_config = &dev->ipc_config;
+	union ipc4_notification_header *primary =
+		(union ipc4_notification_header *)&msg_proto.header;
+	struct ipc_msg *msg;
+	struct tdfb_notification_payload *payload;
+
+	memset_s(&msg_proto, sizeof(msg_proto), 0, sizeof(msg_proto));
+	primary->r.notif_type = SOF_IPC4_MODULE_NOTIFICATION;
+	primary->r.type = SOF_IPC4_GLB_NOTIFICATION;
+	primary->r.rsp = SOF_IPC4_MESSAGE_DIR_MSG_REQUEST;
+	primary->r.msg_tgt = SOF_IPC4_MESSAGE_TARGET_FW_GEN_MSG;
+	msg = ipc_msg_w_ext_init(msg_proto.header, msg_proto.extension,
```

Comment:
`msg_proto.extension` hasn't been set, so it's 0, right? Could use it explicitly to make it clearer